cpp/src/arrow/engine/substrait/expression_internal.cc

```diff
@@ -702,61 +811,141 @@ struct ScalarToProtoImpl {
     return Status::OK();
   }
 
+  Status EncodeUserDefined(const DataType& data_type,
+                           const google::protobuf::Message& value) {
+    ARROW_ASSIGN_OR_RAISE(auto anchor, ext_set_->EncodeType(data_type));
+    auto user_defined = std::make_unique<Lit::UserDefined>();
+    user_defined->set_type_reference(anchor);
+    auto value_any = std::make_unique<google::protobuf::Any>();
+    value_any->PackFrom(value);
+    user_defined->set_allocated_value(value_any.release());
+    lit_->set_allocated_user_defined(user_defined.release());
+    return Status::OK();
+  }
+
+  Status Visit(const NullScalar& s) {
+    ARROW_ASSIGN_OR_RAISE(auto anchor, ext_set_->EncodeType(*s.type));
+    auto user_defined = std::make_unique<Lit::UserDefined>();
+    user_defined->set_type_reference(anchor);
+    lit_->set_allocated_user_defined(user_defined.release());
+    return Status::OK();
+  }
   Status Visit(const BooleanScalar& s) { return Primitive(&Lit::set_boolean, s); }
 
   Status Visit(const Int8Scalar& s) { return Primitive(&Lit::set_i8, s); }
   Status Visit(const Int16Scalar& s) { return Primitive(&Lit::set_i16, s); }
   Status Visit(const Int32Scalar& s) { return Primitive(&Lit::set_i32, s); }
   Status Visit(const Int64Scalar& s) { return Primitive(&Lit::set_i64, s); }
 
-  Status Visit(const UInt8Scalar& s) { return NotImplemented(s); }
-  Status Visit(const UInt16Scalar& s) { return NotImplemented(s); }
-  Status Visit(const UInt32Scalar& s) { return NotImplemented(s); }
-  Status Visit(const UInt64Scalar& s) { return NotImplemented(s); }
-
-  Status Visit(const HalfFloatScalar& s) { return NotImplemented(s); }
+  Status Visit(const UInt8Scalar& s) {
+    google::protobuf::UInt64Value value;
+    value.set_value(s.value);
+    return EncodeUserDefined(*s.type, value);
+  }
+  Status Visit(const UInt16Scalar& s) {
+    google::protobuf::UInt64Value value;
+    value.set_value(s.value);
+    return EncodeUserDefined(*s.type, value);
+  }
+  Status Visit(const UInt32Scalar& s) {
+    google::protobuf::UInt64Value value;
```

Comment:
In protobuf all unsigned integers are encoded the same way and there is no uint16 or uint8 so I figured it would be simplest and most consistent to just use uint64 for everything.  It's something of an arbitrary decision and, even if the receiver decodes it as a uint32, it will still work.